GoogleDataTransport/GDTCORLibrary/GDTCORFlatFileStorageIterator.m

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#import "GDTCORLibrary/Private/GDTCORFlatFileStorageIterator.h"
+
+#import <GoogleDataTransport/GDTCORConsoleLogger.h>
+#import <GoogleDataTransport/GDTCOREvent.h>
+
+@implementation GDTCORFlatFileStorageIterator {
+  /** The current eventFiles array index, incremented in -nextEvent. */
+  NSInteger _currentIndex;
+}
+
+- (instancetype)initWithTarget:(GDTCORTarget)target queue:(dispatch_queue_t)queue {
+  self = [super init];
+  if (self) {
+    _queue = queue;
+    _target = target;
+  }
+  return self;
+}
+
+- (nullable GDTCOREvent *)nextEvent {
+  if (_currentIndex == -1) {
+    return nil;
+  }
+  if (!_eventFiles) {
+    GDTCORLogDebug(@"%@", @"eventFiles property not set, so -nextEvent will be nil.");
+    return nil;
+  }
+  dispatch_queue_t queue = _queue;
+  if (!queue) {
+    GDTCORLogDebug(@"%@", @"iterator queue was nil, so -nextEvent will be nil.");
+    return nil;
+  }
+  __block GDTCOREvent *nextEvent;
+  dispatch_sync(queue, ^{
+    NSData *data = [NSData dataWithContentsOfFile:self->_eventFiles[self->_currentIndex]];
```

Comment:
It looks like it is possible that the file may become unavailable (deleted) by the time it is requested. Is it intended?

Reply:
Yes, this is a good point. I'm not sure how I'm going to address it yet. I have several options:
- check for file exists in -nextEvent
- introduce API to 'lock' files to prevent them from showing up in iterators
- defer finding file paths until -nextEvent (as you suggested), though this may not work because -nextEvent doesn't run on (and block) the storage queue, so the same problem could still occur.